### PR TITLE
feat(fiatconnect): Add KYC checking when selecting quotes

### PR DESCRIPTION
**Pull request by jophish**
### Description

For [ACT-381](https://linear.app/valora/issue/ACT-381/wallet-handle-kyc-in-quoteonpress). Implements the logic coming off of the "Call GET /KYC on IHL" node in [this flowchart](https://www.figma.com/file/fqLxcOIvDxQWAlDY60rCD6/FiatConnect-Persona-Diagram?node-id=12%3A397). In simpler terms, this PR handles all the KYC logic required when selecting a quote on the Select Provider screen.

New `KycStatus`/`KycLanding` screens (they're the same):

2022-09-14-120453_673x1418_scrot

### Other changes

Adds some functionality to make FC-authenticated calls to IHL, turns on support for KYC in quotes.

### Tested

These changes are unit tested and were tested out in the app by mocking out calls to IHL. A full e2e test, actually calling IHL is not currently possible since IHL does not have all the necessary endpoints yet.

### How others should test

_Does this need to be tested by QA in the next release cycle? If so please give a brief explanation of how to test these changes._

### Related issues

- Fixes ACT-381
### Backwards compatibility

_Brief explanation of why these changes are/are not backwards compatible._

### CODE REVIEW

File: src/fiatconnect/KycStatusScreen.tsx [hunk 1]
```diff
@@ -0,0 +1,16 @@
+import * as React from 'react'
```
Comment (jophish) [inline src/fiatconnect/KycStatusScreen.tsx:1]: Added this in as a placeholder to have a unique place to navigate to from the quote handler saga.

File: src/fiatconnect/saga.ts [hunk 2]
```diff
@@ -277,14 +280,57 @@ export function* handleSelectFiatConnectQuote({
   payload: params,
 }: ReturnType<typeof selectFiatConnectQuote>) {
   const { quote } = params
+
   try {
+    // If KYC is required for the quote, check that the user has correct KYC on file
+    // with the quote's provider
+    let getKycStatusResponse: GetKycStatusResponse
+    if (quote.requiresKyc()) {
+      const kycSchema = quote.getKycSchema()!
```
Comment (cajubelt) [inline src/fiatconnect/saga.ts:289]: nit: by using this assertion operator, we are baking in the assumption that if `quote.requiresKyc()` evaluates to true, this will be nonnull. Equivalently, we could make use of the logical contrapositive, e.g. if `quote.getKycSchema()` is falsy, kyc is not required. (And ditch the usage of `quote.requiresKyc()` plus the assertion operator type hack.) This seems preferable to me.

Comment (cajubelt) [reply src/fiatconnect/saga.ts:289]: maybe a rename like `getRequiredKycSchema` could make this suggestion more palatable (since it's fairly clear to me that if there isn't a required kyc schema, I don't have to do kyc)

File: src/fiatconnect/saga.ts [hunk 3]
```diff
@@ -277,14 +280,57 @@ export function* handleSelectFiatConnectQuote({
...
+        case FiatConnectKycStatus.KycPending:
+          break
+        // If denied or expired, skip account management and
+        // navigate to the KYC status screen
+        case FiatConnectKycStatus.KycDenied:
+        case FiatConnectKycStatus.KycExpired:
+          navigate(Screens.KycStatusScreen)
+          yield put(selectFiatConnectQuoteCompleted())
+          return
+      }
```
Comment (cajubelt) [inline src/fiatconnect/saga.ts:323]: might be helpful to be defensive and handle the case where `fiatConnectKycStatus` is undefined or unrecognized

Comment (cajubelt) [reply src/fiatconnect/saga.ts:323]: and log an error or something

File: src/fiatconnect/saga.ts [hunk 4]
```diff
@@ -306,18 +352,28 @@ export function* handleSelectFiatConnectQuote({
...
-    navigate(Screens.FiatConnectReview, {
-      flow: quote.flow,
-      normalizedQuote: quote,
-      fiatAccount,
-    })
+
+    // If the quote required KYC, only proceed to the Review screen if it's approved
+    if (
+      quote.requiresKyc() &&
+      getKycStatusResponse!.kycStatus[quote.getKycSchema()!] !== FiatConnectKycStatus.KycApproved
```
Comment (cajubelt) [inline src/fiatconnect/saga.ts:359]:
```suggestion
      getKycStatusResponse?.kycStatus[quote.getKycSchema()!] !== FiatConnectKycStatus.KycApproved
```

Comment (cajubelt) [reply src/fiatconnect/saga.ts:359]: also, same thing here re `requiresKyc` and `quote.getKycSchema()!`

File: src/fiatconnect/saga.ts [hunk 5]
```diff
@@ -277,14 +280,57 @@ export function* handleSelectFiatConnectQuote({
   payload: params,
 }: ReturnType<typeof selectFiatConnectQuote>) {
   const { quote } = params
+
```
Comment (cajubelt) [inline src/fiatconnect/saga.ts:283]: it seems like this saga does something much different on quote selection when kyc is required vs when it is not. However, after the user has gone through kyc, it seems as though we'll want the other stuff to still happen. Do you have a plan for that?

Comment (cajubelt) [reply src/fiatconnect/saga.ts:283]: seems like a naive approach would be to create a new saga that is "everything after kyc" and invoke that (somehow) when the user has completed submitting their kyc info. 

Another (still fairly naive) would be to simulate selecting the quote again once the user has submitted kyc info

**Review by cajubelt — COMMENTED** [review]
skimmed everything except the saga

File: src/fiatconnect/saga.ts (hunk 2 above)
Comment (jophish) [reply src/fiatconnect/saga.ts:289]: Yeah, I went back and forth between just having `getKycSchema` and using its falsiness as a proxy for knowing whether or not KYC is required, or using both `requiresKyc` and `getKycSchema`. I think using just `getKycSchema is fine; it will definitely make things cleaner.

File: src/fiatconnect/saga.ts (hunk 5 above)
Comment (jophish) [reply src/fiatconnect/saga.ts:283]: Hm, you mean after they've completed the Persona flow in particular, we would want to "resume" with the remainder of the saga (basically run the whole saga as it was before this PR) as normal? Simulating selecting the quote again (just dispatching the same action we do when a user selects a quote) isn't the worst idea imo, it's the most "robust" in that it actually ensures that the KYC states stored in IHL are as we expect, rather than just assuming they are after completing Persona KYC.

Also, after sending KYC to Persona, we'll need to hit POST /kyc for the provider they've selected; rather than doing that separately as an ad-hoc thing, that can just be done when calling this saga again, since the logic to do that already exists there. If we go this route, perhaps the naming of this saga could use an update, though.

File: src/fiatconnect/KycStatusScreen.tsx [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+import { StyleSheet } from 'react-native'
+import { SafeAreaView } from 'react-native-safe-area-context'
+
+function KycStatusScreen() {
```
[the remark this answers is not in the document]
Comment (jophish) [reply src/fiatconnect/KycStatusScreen.tsx:5]: Added

**Review by cajubelt — APPROVED** [review]

File: src/in-house-liquidity/index.test.ts [hunk 7]
```diff
@@ -54,4 +65,217 @@ describe('In House Liquidity Calls', () => {
       ).toThrow('Cannot call IHL because walletAddress is null')
     })
   })
+
+  describe('makeRequest', () => {
+    it('makes a FC authenticated call', async () => {
+      ;(getAuthHeaders as jest.Mock).mockResolvedValueOnce({ 'header-key': 'header-val' })
```
Comment (satish-ravi) [inline src/in-house-liquidity/index.test.ts:71]: you can used [jest.mocked(...)](https://jestjs.io/docs/mock-function-api/#jestmockedsource-options) instead of `as`. Or actually `mocked` helper from `ts-utils/jest` since we are in an older version of jest

File: src/in-house-liquidity/index.test.ts [hunk 8]
```diff
@@ -54,4 +65,217 @@ describe('In House Liquidity Calls', () => {
       ).toThrow('Cannot call IHL because walletAddress is null')
     })
   })
+
+  describe('makeRequest', () => {
+    it('makes a FC authenticated call', async () => {
```
Comment (satish-ravi) [inline src/in-house-liquidity/index.test.ts:70]: Can you also include `if provider info is passed / not passed` to the test name? Wasn't super obvious when reading these tests.
Also, should these tests also assert that the function returns correct response?

File: src/in-house-liquidity/index.test.ts [hunk 9]
```diff
@@ -54,4 +65,217 @@ describe('In House Liquidity Calls', () => {
...
+        method: 'GET',
+      })
+    })
+  })
+
+  describe('getAuthHeaders', () => {
+    let mockFiatConnectClient: jest.Mocked<FiatConnectApiClient>
+
+    beforeEach(async () => {
+      mockFiatConnectClient = (await getFiatConnectClient(
```
Comment (satish-ravi) [inline src/in-house-liquidity/index.test.ts:113]: any reason to call getFiatConnectClient here? can't we just use the mock functions we define above as global vars?

File: src/navigator/Navigator.tsx [hunk 10]
```diff
@@ -511,6 +513,12 @@ const settingsScreens = (Navigator: typeof Stack) => (
       name={Screens.FiatConnectLinkAccount}
       component={FiatConnectLinkAccountScreen}
     />
+    <Navigator.Screen options={headerWithBackButton} name={Screens.Persona} component={Persona} />
```
Comment (satish-ravi) [inline src/navigator/Navigator.tsx:516]: I don't think the Persona component is a screen. It's currently just a button, which on clicking opens the persona flow. Is the plan to refactor this component to just load Persona flow on load? Or do we just want to refactor the persona stuff into it's own saga and drop that component?

File: src/fiatconnect/saga.ts [hunk 11]
```diff
@@ -277,14 +280,60 @@ export function* handleSelectFiatConnectQuote({
...
+          }
+        // If approved or pending, continue as normal and handle account management
+        case FiatConnectKycStatus.KycApproved:
+        case FiatConnectKycStatus.KycPending:
+          break
+        // If denied or expired, skip account management and
+        // navigate to the KYC status screen
+        case FiatConnectKycStatus.KycDenied:
+        case FiatConnectKycStatus.KycExpired:
+          navigate(Screens.KycStatusScreen)
```
Comment (satish-ravi) [inline src/fiatconnect/saga.ts:320]: This screen should eventually take flow, quote, and kyc status as props right?

File: src/in-house-liquidity/index.test.ts (hunk 9 above)
Comment (jophish) [reply src/in-house-liquidity/index.test.ts:113]: I thought that calls to jest.mock were hoisted so you couldn't reference global vars within them?

File: src/fiatconnect/saga.ts (hunk 11 above)
Comment (jophish) [reply src/fiatconnect/saga.ts:320]: Yep, that's probably the case.

File: src/navigator/Navigator.tsx (hunk 10 above)
Comment (jophish) [reply src/navigator/Navigator.tsx:516]: Added a new `KycLanding` screen instead!

File: src/in-house-liquidity/index.test.ts (hunk 7 above)
Comment (jophish) [reply src/in-house-liquidity/index.test.ts:71]: This introduces another issue with TS complaining about resolved result type :frowning_face:

**Review by satish-ravi — APPROVED** [review]

File: src/in-house-liquidity/index.test.ts (hunk 9 above)
Comment (satish-ravi) [reply src/in-house-liquidity/index.test.ts:113]: I think this works if global variables are referenced inside another jest.fn(). E.g., https://github.com/valora-inc/wallet/pull/2855/files#diff-7caff590ac25eb7fe43d613aa98c7e1b1d4e472211ab028f2aa17b11b2514bc3R39

**Comment by ValoraQA** [comment]
Hey @MuckT as discussed in weekly meet ,we are skipping this PR 

Thanks!
